llama.cpp

```diff
@@ -11861,7 +11861,9 @@ static void llama_grammar_advance_stack(
         std::vector<std::vector<const llama_grammar_element *>> & new_stacks) {
 
     if (stack.empty()) {
-        new_stacks.emplace_back(stack);
+        if (std::find(new_stacks.begin(), new_stacks.end(), stack) == new_stacks.end()) {
```

Comment:
Would it be worth using an std::set? (or unordered_set maybe?)

Reply:
Scratch that, I tried an std::set and binary search / insert at sorted place in vector, doesn't make a difference with json_numbers.grammar.

Can you share the diff of you do the timing in test-grammar-integration?

Reply:
That's a great question -- I looked at it a bit, but I'm on the fence. std::set would change this lookup operation from O(N) to O(log(N)), but it also has other overhead related to maintaining ordering and whatnot (that we don't care about). Is a thinner O(N) operation cheaper than a thicker O(log(N))? So I wasn't sure if the additional costs would be worth the gains -- I guess eventually this would be worth testing, but might be worth saving for a future PR?